rfcs/0062-content-addressed-paths.md

```diff
@@ -60,22 +60,75 @@ cutoffs.
 In all that follows, we pretend that each derivation has only one output.
 This doesn't change the reasoning but makes things easier to state.
 
-The gist of the design is that
+The gist of the design is that:
 
 - Some derivations can be marked as content-adressed (ca), in which case their
   output will be moved to a path `ca` determined only by its content after the
   build
-- Each (non content-adressed) derivation will have two outputs: A `static` one
-  computed at evaluation time and a `dynamic` one computed from the dynamic
-  outputs of its dependencies. These outputs may be identical if the derivation
-  doesn't (transitively) depend on any ca derivation
-- just prior to being realized, each derivation gets rewritten by replacing
-  each of its dependencies by its `dynamic` or `ca` path
+- We introduce the notion of a `resolved derivation` which is a derivation that
+  doesn't refer to any other derivation but only to concrete store paths.
+  To prevent ambiguities, we might speak of a `symbolic derivation` to
+  designate a derivation that's not necessarily resolved.
+  We also define a `resolving` function that given a symbolic derivation
+  returns a new resolved derivation with the same semantics.
+- When asked to build a derivation, Nix will first resolve it, build the
+  resolved derivation and link back the symbolic one to the out path of the
+  resolved one.
 
-## Example
+## Nix-build process
+
+### Output mappings
+
+A major consequence of allowing content-addressed derivations is that the
+actual output path of a derivation might not match its output hash anymore.
+
+To express this, we introduce a new mapping `pathOf` that associates the hash
+of every live derivation to its store path.
+By extension, we also define `pathOf(drv) = pathOf(hash(drv))`
+
+### Building a ca derivation
+
+ca derivations are derivations with the `__contentAddressed` argument set to
+`true`.
+
+The process for building a content-adressed derivation is the following:
+
+- We build it like a normal derivation (see below) to get an output path `$out`.
+- We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
+- We move `$out` to `/nix/store/$chash-$name`
+- We create a mapping from `$dhash` (the hash computed at eval-time) to
+  `/nix/store/$chash-$name`
+
+[^modulo-hashing]:
+
+  We can possibly normalize all the self-references before
+  computing the hash and rewrite them when moving the path to handle paths with
+  self-references, but this isn't strictly required for a first iteration
```

Comment:
I would very much like to *not* do this the first drift.

Reply:
Nix already supports this, behind a flag.

Reply:
It does, but I'd still like to be a bit cautious. When the mounted store path is the content addressed thing, we open the door to all sorts of interesting ways to distribute and deploy nix stores. By adding this post-processing step we complicate those at the very least.